File: app/src/main/java/com/aar/app/wordsearch/di/modules/AppModule.java
package com.aar.app.wordsearch.di.modules;

import android.app.Application;
import android.content.Context;
import android.content.SharedPreferences;
import android.preference.PreferenceManager;

import com.aar.app.wordsearch.features.ViewModelFactory;
import com.aar.app.wordsearch.data.GameDataSource;
import com.aar.app.wordsearch.data.GameThemeRepository;
import com.aar.app.wordsearch.data.WordDataSource;
import com.aar.app.wordsearch.features.gamehistory.GameHistoryViewModel;
import com.aar.app.wordsearch.features.gameover.GameOverViewModel;
import com.aar.app.wordsearch.features.gameplay.GamePlayViewModel;
import com.aar.app.wordsearch.features.mainmenu.MainMenuViewModel;

import javax.inject.Singleton;

import dagger.Module;
import dagger.Provides;

/**
 * Created by abdularis on 18/07/17.
 */

@Module
public class AppModule {

    private Application mApp;

    public AppModule(Application application) {
        mApp = application;
    }

    @Provides
    @Singleton
    Context provideContext() {
        return mApp;
    }

    @Provides
    @Singleton
    SharedPreferences provideSharedPreferences(Context context) {
        return PreferenceManager.getDefaultSharedPreferences(context);
    }

    @Provides
    @Singleton
    ViewModelFactory provideViewModelFactory(GameDataSource gameDataSource,
                                             WordDataSource wordDataSource) {
        return new ViewModelFactory(
                new GameOverViewModel(gameDataSource),
                new GamePlayViewModel(gameDataSource, wordDataSource),
                new MainMenuViewModel(new GameThemeRepository()),
                new GameHistoryViewModel(gameDataSource)
        );
    }
}

File: app/src/main/java/co/africanwolf/hiddenwords/di/AppModule.kt
package co.africanwolf.hiddenwords.di

import android.content.Context
import android.content.Context.MODE_PRIVATE
import android.content.SharedPreferences
import co.africanwolf.hiddenwords.ViewModelFactory
import co.africanwolf.hiddenwords.data.GameThemeRepository
import co.africanwolf.hiddenwords.data.WordDataSource
import co.africanwolf.hiddenwords.data.sqlite.DbHelper
import co.africanwolf.hiddenwords.data.sqlite.GameDataSQLiteDataSource
import co.africanwolf.hiddenwords.gamehistory.GameHistoryViewModel
import co.africanwolf.hiddenwords.gameover.GameOverViewModel
import co.africanwolf.hiddenwords.gameplay.GamePlayViewModel
import co.africanwolf.hiddenwords.mainmenu.MainMenuViewModel
import dagger.Module
import dagger.Provides
import dagger.hilt.InstallIn
import dagger.hilt.android.qualifiers.ApplicationContext
import dagger.hilt.components.SingletonComponent
import javax.inject.Singleton

@Module
@InstallIn(SingletonComponent::class)
object AppModule {

    @Provides
    @Singleton
    fun provideSharedPreferences(@ApplicationContext context: Context): SharedPreferences {
        return context.getSharedPreferences("word-search", MODE_PRIVATE)
    }

    @Provides
    @Singleton
    fun provideViewModelFactory(
        gameDataSource: GameDataSQLiteDataSource,
        wordDataSource: WordDataSource
    ): ViewModelFactory {
        return ViewModelFactory(
            GameOverViewModel(gameDataSource),
            GamePlayViewModel(gameDataSource, wordDataSource),
            MainMenuViewModel(GameThemeRepository()),
            GameHistoryViewModel(gameDataSource)
        )
    }

    @Provides
    @Singleton
    fun provideDbHelper(@ApplicationContext context: Context?): DbHelper {
        return DbHelper(context)
    }
}
